tests/013-eventhandler.cpp

```diff
@@ -0,0 +1,148 @@
+#include <catch2/catch_test_macros.hpp>
+#include <catch2/matchers/catch_matchers.hpp>
+#include <future>
+#include <oxen/quic.hpp>
+#include <oxen/quic/gnutls_crypto.hpp>
+#include <thread>
+
+#include "utils.hpp"
+
+namespace oxen::quic::test
+{
+    struct lifetime : public std::enable_shared_from_this<lifetime>
+    {};
+
+    TEST_CASE("013 - EventHandler event repeater: calling object lifetime bound", "[013][repeater][caller]")
+    {
+        Network test_net{};
+        const int NUM_ITERATIONS{10};
+        constexpr auto msg = "hello from the other siiiii-iiiiide"_bsv;
+
+        std::promise<bool> d_promise;
+        std::future<bool> d_future = d_promise.get_future();
+
+        std::atomic<int> recv_counter{}, send_counter{};
+
+        stream_data_callback server_data_cb = [&](Stream&, bstring_view) { recv_counter += 1; };
+
+        auto [client_tls, server_tls] = defaults::tls_creds_from_ed_keys();
+
+        Address server_local{};
+        Address client_local{};
+
+        auto server_endpoint = test_net.endpoint(server_local);
+        REQUIRE_NOTHROW(server_endpoint->listen(server_tls, server_data_cb));
+
+        RemoteAddress client_remote{defaults::SERVER_PUBKEY, "127.0.0.1"s, server_endpoint->local().port()};
+
+        auto client_endpoint = test_net.endpoint(client_local);
+        auto conn_interface = client_endpoint->connect(client_remote, client_tls);
+
+        // client make stream and send; message displayed by server_data_cb
+        auto client_stream = conn_interface->open_stream();
+
+        auto life = std::make_shared<lifetime>();
+
+        test_net.call_every(10ms, life->weak_from_this(), [&]() {
+            if (send_counter <= NUM_ITERATIONS)
+            {
+                ++send_counter;
+                client_stream->send(msg);
+            }
+            else
+                life.reset();
+        });
+
+        test_net.call_later(1s, [&]() {
+            REQUIRE(recv_counter == send_counter);
+            REQUIRE(!life);
```

Comment:
Invoking REQUIRE in another thread like this is a recipe for crashes with catch2; you *can* make it work if you use a mutex to guard anything that touches catch2, but it would be easier to get these values out via a promise and then REQUIRE after the promise gets fulfilled.

Reply:
(And there's a `d_promise` just below this that can do the job)

Reply:
I understand your motivation behind this. Few thoughts on that:
- By the time the `::call_later()` is invoked, the sending and receiving has stopped
- The counters are atomic
- There are literally tons of other non-problematic `REQUIRE` checks in the test code that aren't guarded my mutexes

Reply:
Not resolved.

`REQUIRE` inside a callback is something we should avoid if easy to do so precisely because of the Catch2 thread hazard; even if this call does not hit the problem, it is also extremely simple to move the `REQUIRE`s after the `require_future` without sacrificing anything in terms of the test or clarity, but gaining that it is immediately obviously safe since there's no longer any catch2 interactions happening on different threads in the test.